### PR TITLE
fix(dialog,popover): restore focus when `open` set to `false`

### PR DESCRIPTION
**Pull request by anatolzak**
closes https://github.com/melt-ui/melt-ui/issues/1109

We now restore focus to the trigger (or using the `closeFocus` prop) in an `effect` that runs when `open` is `false`. In order to not focus the trigger on initial mount when `open` is `false`, I added a parameter `skipFirstRun` to the `effect` function that optionally skips the first run of the effect.

I also added regression tests for the dialog and popover. 

Also, in the menu builder, we are already doing what I implemented with the dialog and popover, which removes the need in the menu builder to manually focus the trigger in multiple places. And I cleaned up the effect in the menu component. 

### Dialog Before
https://github.com/melt-ui/melt-ui/assets/53095479/584dd6ad-bd39-4901-9ff1-2a0e987f1a2e

### Dialog After 
https://github.com/melt-ui/melt-ui/assets/53095479/ac7ac1d1-f02e-4b2a-8f48-040438149a43

### Popover Before

https://github.com/melt-ui/melt-ui/assets/53095479/89b2dc56-e1a8-47e3-9ce3-d4c42fbe41c3

### Popover After

https://github.com/melt-ui/melt-ui/assets/53095479/fa3943a1-35d1-417b-afe9-cb4f5650b69d

### CODE REVIEW

File: src/lib/internal/helpers/store/effect.ts [hunk 1]
```diff
@@ -10,18 +10,25 @@ import { safeOnDestroy } from '../lifecycle.js';
...
  * @param stores - The stores object to derive from
  * @param fn - The function to run when the stores change
+ * @param skipFirstRun - Whether to skip the first run
  * @returns A function that can be used to unsubscribe the effect
  */
 export function effect<S extends Stores>(
 	stores: S,
-	fn: (values: StoresValues<S>) => (() => void) | void
+	fn: (values: StoresValues<S>) => (() => void) | void,
+	skipFirstRun?: boolean
```
Comment (huntabyte) [inline src/lib/internal/helpers/store/effect.ts:19]: Let's make this an object. @TGlide and I have spoke in the past about some other options we want to add to effects and by making it an object now it'll ease the addition of those.

```ts
type EffectOptions = {
  skipFirstRun?: boolean;
}
```

**Review by huntabyte — CHANGES_REQUESTED** [review]
Minor thing! Thanks man!

**Comment by anatolzak** [comment]
@huntabyte I added an options parameter to `effect` as you requested!

**Review by huntabyte — APPROVED** [review]
Thank you!

**Pull request merged**